### ISSUE TITLE
How to make this work with SSR?

### ISSUE THREAD

**Issue by JasonStoltz** [elastic/search-ui#78]

**Comment by JasonStoltz** [comment]
Relates to:
https://github.com/elastic/search-ui/issues/77

**Comment by FokkeZB** [comment]
I wish I had seen this issue earlier 🤦‍♂ 

We're trying to use `search-ui` for the new Help Center at zapier.com, which uses NextJS. It works fine for the most part, but what I struggle with and can't quite put my finger on is my root `WithSearch` (directly under `SearchProvider` and `ErrorBoundary` and those wrapping the `search-ui` container components getting out of sync. On the initial (server-side) render the root sees the query but further down they don't. And when I change the query cient-side (using `setSearchTerm()` the old results show instead of the new:

![](https://zappy.zapier.com/029B6527-0DE4-4509-AA86-CDB4F25FAE5D.gif)

I've tried, with no luck:

- Changing `WithSearch` from `PureComponent` to `Component`, thinking that may stop the state change to flow down (although AFAIK `PureComponent` doesn't apply to any child components).
- Not using `setSearchTerm()` but leave `trackUrlState: true` and let `search-ui` manage and listen to the query string.
- Using `SearchContext.Consumer`, which **did** solve for when the page initially renders, but when I change the query it doesn't display anything (not even the old), probably because of a shallow comparison of the context doesn't trigger re-renders as the `driver` object is still the same.
- Reproducing the issue in the (non-SSR demo) https://codesandbox.io/s/search-ui-national-parks-example-3qb66

Next on my list is to not use any of the container components (we weren't using the views anyway) and just take everything what I need via `WithSearch` and pass that down to our views.

But any suggestions or help is more than welcome :)

**Comment by JasonStoltz** [comment]
Hey @FokkeZB. It's exciting to hear that you're using search-ui for the zapier.com help desk. I can definitely help you work through this issue.

As for SSR in general, I'm actually unsure of exactly what user expectations would be for a project like this in SSR. What I ended up doing was moving the initialization of `SearchProvider` to `componentDidMount`, to avoid rendering server side all together.

I can't say I've seen anything like this happen before, and it's a bit hard to debug without seeing any code or being able to play with it in a browser.

> Not using setSearchTerm() but leave trackUrlState: true and let search-ui manage and listen to the query string.

Do you have `trackUrlState` set to `false`?

> Using SearchContext.Consumer, which did solve for when the page initially renders, but when I change the query it doesn't display anything (not even the old), probably because of a shallow comparison of the context doesn't trigger re-renders as the driver object is still the same.

This sounds concerning.

I can't make any suggestions, unfortunately, without being able to reproduce the issue. I'll see if I can spin something up in Next and reproduce this.

**Comment by FokkeZB** [comment]
Yes! I managed to reproduce it in the sandbox and it's not related to SSR at all, so I've created a new issue for it: #279

We're looking to launch our new help center pretty soon, so I'm super grateful for your quick responses. Thanks!

**Comment by FokkeZB** [comment]
About this:

> What I ended up doing was moving the initialization of SearchProvider to componentDidMount, to avoid rendering server side all together.

How exactly did you do that? I was planning to pass `isServerSide: !!context.req` from `getInitialProps()` and then render `{isServerSide && <SearchProvider ... />}`.

**Comment by JasonStoltz** [comment]
@FokkeZB Awesome, looking now.

**Comment by JasonStoltz** [comment]
@FokkeZB 
Here: https://github.com/elastic/search-ui/blob/master/packages/react-search-ui/src/SearchProvider.js#L31

And here: https://github.com/elastic/search-ui/blob/master/packages/react-search-ui/src/SearchProvider.js#L46

**Comment by FokkeZB** [comment]
Ah, so `SearchProvider` takes care of that, awesome!

**Comment by JasonStoltz** [comment]
I'm going to close this issue, as it's documented how this works with SSR, basically. Will open a new issue if there is every a specific need for this to work differently.

**Comment by FokkeZB** [comment]
Well, after #280 ;)
